### ISSUE TITLE
[css-contain-3] Revisit decision to make `style` the default `container-type`

### ISSUE THREAD

**Issue by mirisuzanne** [w3c/csswg-drafts#7066]
In https://github.com/w3c/csswg-drafts/issues/6393#issuecomment-1011283604 we resolved that:

> Boxes default to style containment

@una reopened that discussion in https://github.com/w3c/csswg-drafts/issues/6393#issuecomment-1012446872, and it was briefly discussed again in https://github.com/w3c/csswg-drafts/issues/6644#issuecomment-1034036418 without any definitive conclusion. 

It seemed like a good idea for this discussion to have its own issue. 

The primary concerns on each side are:
- Keeping the `style` default makes one container type inconsistent with all other types. And while it helps with queries on inherited properties, it makes `container-name` much more important for style queries on non-inherited properties (like background-color).
- Removing the style default makes it much harder to query inherited properties, which should almost always resolve against the direct parent. Without the default value, authors are likely to add the `style` container-type to all elements with the universal selector, which is less performant.

Related, I'm curious about the details of what it means to have `style` as a default. Do we mean…
- …that `style` is the initial value of the property, and any other value overrides that behavior? In this case, authors would have to write e.g. `container-type: size style` in order to keep the default intact while adding other container types. That doesn't seem ideal.
- …that all elements are style containers, no matter what value is applied to the property? In this case, we might not need an explicit `style` value? Or a variant might be that the default can be removed with `none` and re-applied with `style`?

**Comment by fantasai** [comment]
I think the reasonable answer is “that all elements are style containers, no matter what value is applied to the property”, otherwise authors will break things accidentally as they adopt size queries throughout their pages. That said, we could reasonably add a value that removes the ability to query at all, as long as it's an explicit disabling value, like `no-style` or something.

**Comment by astearns** [comment]
I’d like to see more discussion here before we spend time on the call again. @una do you have any thoughts on this since the last time?

**Comment by emilio** [comment]
I'm a bit concerned about making `style` the initial container type. It can have somewhat annoying performance implications in the sense that now even non-inherited style changes can affect styles in the elements' subtree.

It's not impossible to optimize some of the cases, but it seems unfortunate to add overhead by default.

**Comment by mirisuzanne** [comment]
@emilio It seems likely to me that if we don't make `style` the default, authors are likely to add it as a global setting anyway - similar to the way they currently change `box-sizing` globally. Without that, it is very difficult to use style-queries with inherited properties in any practical way. That seems like an even worse option for performance in my mind?

**Comment by emilio** [comment]
Maybe? Not sure about that, I'm concerned about is the defaults... If you opt into everything being a container, then sure, so be it, but you've done it explicitly. But making the default behavior more expensive seems a bit unfortunate.

**Comment by anttijk** [comment]
I'm strongly opposed to style (or any) containers being the default for the reasons mentioned my Emilio.

**Comment by matthew-dean** [comment]
@mirisuzanne 

>  It seems likely to me that if we don't make style the default, authors are likely to add it as a global setting anyway - similar to the way they currently change box-sizing globally. 

I was going to ask this in a different thread but I found this here. It seems to me that the way the spec is currently defined, it will soon become "best practice" to do something like:

```css
html {
  box-sizing: border-box;
}
*, *::before, *::after {
  box-sizing: inherit;
  container-type: style size;
}
```
That is, I can't think of a scenario where I wouldn't want the ability to drop in a container query for size or style whereever I please. But, that said, I don't really understand the performance implications of setting this globally but _not_ querying the style.

**Comment by mirisuzanne** [comment]
@matthew-dean I strongly agree that it will become best-practice to add `style` container-type globally (tho the current spec does that by default). However, authors will need to be much more careful about applying `size` containers, because the added size containment removes all intrinsic sizing of children from the container's intrinsic size. Every element would layout as though empty if you applied that globally. Since a lot of CSS layout relies on containers using the intrinsic sizing of their child elements, that's a pretty dangerous tradeoff that authors will need to consider case-by-case.

**Comment by matthew-dean** [comment]
@mirisuzanne 

> However, authors will need to be much more careful about applying size containers, because the added size containment removes all intrinsic sizing of children from the container's intrinsic size. Every element would layout as though empty if you applied that globally. Since a lot of CSS layout relies on containers using the intrinsic sizing of their child elements, that's a pretty dangerous tradeoff that authors will need to consider case-by-case.

Oh, interesting. Is there an explainer for this that demonstrates the visual effect of applying different `container-type` properties?

I don't know if this is useful, regarding applying `size` which effectively removes `style`, Less has a syntax for "adding properties" to existing values, as well as a keyword like `all` which applies all items from a possible list. The former (a CSS native way of adding to an inherited list value) may be out of scope for this feature, but the latter (a single keyword representing both) could be a possibility.

**Comment by mirisuzanne** [comment]
> Oh, interesting. Is there an explainer for this that demonstrates the visual effect of applying different `container-type` properties?

@matthew-dean I think this behavior is described in several places, but I don't know if any of those descriptions really 'demonstrate the visual effect' in a simple & clear author-centered way. That would be a good example to have somewhere. We could consider adding a note/example to the spec, but it will be even more helpful to call out in articles about the feature.

> Less has a syntax for "adding values" to existing list values, as well as a keyword like all which applies all items from a possible list.

There are several other open issues about that exact idea, and I think we do need something like that. But I don't see any reason to define it in a way that is specific to container queries. It's something we would want to define generically for all list values, and should be handled elsewhere.

---

@astearns I think it would make sense to bring this issue back to the group? With several implementors arguing against the current behavior, and several browsers preparing to ship the feature, it seems important to discuss the tradeoff they bring up. I think this issue needs to be resolved before any version of the feature can be shipped, since it impacts the default value of the property?

- on the one hand, in order for style queries to be useful on inherited properties, authors will want style containment set globally. If that is an author setting, they will also need to be careful not to override it when adding `size` containers. Since size CQs are likely to launch before style CQs, this will create some unexpected difficulties for authors once both are supported. It's not as simple as adding the global `style` value, it will also have to be added to every instance of `container-type` that might accidentally override that global setting. If we go that rout, we need to clarify that the default style-container behavior is not overridden by author values ([as described by fantasai](https://github.com/w3c/csswg-drafts/issues/7066#issuecomment-1042410631)). 
- on the other hand, it sounds like this desired behavior may cause performance issues for browser engines. 

I think we've made the right choice about how the syntax should ideally work for authors. But, of course, performance is also important to authors. It's not clear to me how those concerns should be balanced, or how bad the performance impacts would be, but I'd like to resolve that question one way or the other so that we don't accidentally paint ourselves into a corner. 

In either case, this issue needs clarification in the spec.

**Comment by lilles** [comment]
I think the current spec with initial value being `style` is the least attractive here because:

- Authors will accidentally stop style queries from working with `container-type: inline-size` as @fantasai [points out](https://github.com/w3c/csswg-drafts/issues/7066#issuecomment-1042410631)
- It means shipping size queries separately will require a different initial value than what's given per spec

Having all elements being style containers by default will require authors to be a bit careful with selectors inside style queries. Queries like this (selectors matching a whole lot of elements) could easily cause implementations to take a performance hit:

```
@container style(--foo: bar) {
  :first-child { color: green }
}
```

**Comment by fantasai** [comment]
@lilles I agree that if we're making style the default, specifying other values shouldn't make it no longer a style container, just like specifying counters to increment doesn't remove the list-item increment from list items.

**Comment by fantasai** [comment]
+1 to @lilles, if `style` is the default we need to make specifying other keywords not override it (same as adding additional counter-resets doesn't drop the `list-item` increment).

**Comment by andruud** [comment]
@fantasai How about we make "additive cascade" happen, and live with `container-type` issues until then? :P

Maybe there's a nice way to split `container-type` into multiple longhands, but if there isn't, it would be sad to add a more general solution later that would make those longhands look stupid in retrospect.

**Comment by lilles** [comment]
It's not clear to me how container-type should be split into different longhands.

If we split into separate longhands, we would still have the issue with the container shorthand?

Say we have a new 'state' container type and split into size-container and state-container properties.

Setting both name and inline-size via the shorthand will reset the state-container:

```
#container {
  state-container: foo
}

#container {
  container: name / inline-size
}
```

If we add:

```
size-container: (none | inline-size | size)
```

Do we want to have a short-hand for setting size-container + container-name?

That is:

```
size-container-type: (none | inline-size | size)
size-container: name / size-container-type
```

Also, would that mean that we would not longer need the full container shorthand. Still an issue with a shorthand with:

```
size-container: size-name / inline-size
state-container: state-name / foo
```

if the first value is container-name

**Comment by andruud** [comment]
Since multiple vendors seem to think it's a little late to get creative here, I propose that we do this:

- `container-type:none` is the initial value, and actually means "none", without implying a style container.
- Authors who prefer style-by-default must then do `* { container-type:style; }`.
- Authors who want to add size-container to something must then remember to do `div { container-type:size style; }` to preserve the style container type. For now.
- Additive cascade improves the situation in the future.

I understand that it's not ideal, but I don't understand why it's necessary to do an emergency hack for this particular case when a) other parts of the CSS has had the same problem all the time (contain, animations, backgrounds, etc), and b) the work-around for the near/medium future is basically "remember to add 'style' to container-type".

Yes, other kinds of containers are vaguely planned, and if that somehow makes the situation unacceptable, then we can focus on doing additive cascade _first_.

**Comment by dbaron** [comment]
Does that mean that when engines start shipping style containment, there will be a performance penalty for existing uses of size containment that don't need style containment?

**Comment by mirisuzanne** [comment]
> How about simply making inline-size and size also be style containers (they already imply style containment)?

This seems like a one-off solution that would protect only the `style` container-type from overrides. The concern that @fantasai has raised is that we may want a more generic solution for any number of future container types that may overlap and conflict in different ways. I get that concern.

But I also agree with @andruud's assertion that _these problems already exist many other places in CSS_. To that point,  `container-name` should also ideally be 'additive' – allowing authors to establish different names in different places and for different reasons, without those names overriding each other. That problem is not easily solved by breaking out more longhands. A container can have multiple size names, just as likely as it has names relating to different container types. I don't think it makes sense to have style-names and size-names and state-names, etc. So it makes sense to me that a 'generic' solution should be _even more generic_ (we need additive cascade), rather than looking for a solution that is generic only to container-types. 

But generic additive cascade is not an easy problem to solve. So, short-term, do the longhands help us get there?

## the longhand approach

Say browsers launch with two longhand properties, and plans for a third (if not more):

```css
/* launching soon… */
size-container: [none | size | inline-size];
container-name: [none | <custom-ident>+];

/* someday later… */
style-container: [none | style];
```

If we stop there, it works. When `style-container` lands, authors can add it as desired, without causing any `size-container` conflicts. But as soon as we add a shorthand, we're back where we started – a single property that overrides future properties:

```
/* launching soon… */
container: <container-name> / <size-container>?;

/* someday later… */
container: <container-name> / <size-container>? <style-container>?;
```

In the immediate, authors will set `container: name / inline-size`, seeing it as a helpful shortcut. But as soon as we add style containers, that new longhand is already in conflict with any use of the existing shorthand. Authors are overriding a value that didn't even exist when they initially wrote the code.

Maybe that's what @fantasai had in mind – longhands, without any shorthand? I think it works, and sidesteps @lilles concerns about merging multiple shorthands. This seems like a viable path to me (tho I still want additive cascade for container-names).

## on the utility of 'style queries' generally

Along the way, several people have questioned if 'style queries' are even useful, especially if we have individual inline functions like `toggle()`/`cycle()` or even `if()`. I'm working on a larger summary of the use-cases here, and will post that soon. But the quick summary is, functions like `toggle()` only work when:

1. you want to set a _single property_
2. based on the parent value of _only that same property_

That works (and is even elegant) if you want to simply cycle a `font-style` between `normal` and `italic`, based on the parent `font-style`. But it falls apart as soon as you get any more complex about the relationships:

1. adding multiple conditions
2. styling a different property than you query
3. styling multiple properties based on a shared condition 

All of those become more essential for use-cases discussed in #5624, but are even required for this simple case (based on a demo from @una elsewhere):

```css
@container style(font-style: italic) {
  em, i, q {
    background: lightpink;
  }
}
```

**Comment by SebastianZ** [comment]
It's a bit late here, so maybe the idea is not completely thought through. Though what about basing querying solely on `container-name` for now and deferring `container-type` and everything about style containers?

That would mean, it's required to specify a name for the container to establish it as a query container. The queries then also require to specify the name additionally to the (size) container query.

This should allow implementations to ship without issues. And for a later version of the specification we can then come back and resolve the issues around style containers and make the name optional again.

Sebastian

**Comment by mirisuzanne** [comment]
@SebastianZ Container type is essential for size containers, even setting aside style for now, because it applies the necessary containment (layout, size/inline-size, style), which can be pretty invasive. What containment is needed depends on the specifics of the type, which we can't infer from a name alone.

**Comment by andruud** [comment]
Looks like an inconsistent mess to me.

**Comment by SebastianZ** [comment]
> @SebastianZ Container type is essential for size containers, even setting aside style for now, because it applies the necessary containment (layout, size/inline-size, style), which can be pretty invasive. What containment is needed depends on the specifics of the type, which we can't infer from a name alone.

So it probably _was_ too late yesterday.😆 The idea would only work if we made size containers the default.

Just to summarize how container queries should work as proposed for my (and hopefully other people's) understanding:

Size containers queries: closest ancestor query container matching the queried container features and optional name
Style containers queries:
* non-inherited properties: closest ancestor query container matching the queried container features and optional name
* inherited properties: parent element (<= That's the main issue.)

Sebastian

**Comment by mirisuzanne** [comment]
Sorry for another long comment, but I feel like there are again several issues getting confused in the thread - and I'd like to provide some context for how they pull apart, and how they relate. (along with the resolution that I would like to see here)

> The idea would only work if we made size containers the default.

Even different size containers rely on different containment (1d or 2d), so there's still no obvious 'default' that we could apply, even with size-alone. There's no way to remove container-type without severely limiting the feature.

> Just to summarize how container queries should work as proposed…

All queries use the same logic: querying the nearest ancestor container that is able to reflect the requested features, optionally with a given name. 

In most cases, for both size and style, it would be ideal to query the direct parent by default - because the baseline purpose of container queries is to have the most immediate information available. 

That's not viable with size queries, because they require invasive containment in order to 'reflect the requested features' – so authors need to carefully select where that containment is applied. With style queries, any element would be 'able to reflect' the query without any intervention, because no containment is required. The same seems likely for the potential 'state' queries that have been discussed so far.

In cases where authors don't just want the nearest ancestor, but want a specific container, they can use a container-name to look farther up the tree. That will be useful across all query types, and is especially a good idea when querying non-inherited properties.

All of this has lead to several distinct questions that are getting mixed together in this thread:

## Are style queries even useful?

I'm confident that they solve a problem that has been raised in various places, and I will continue to make the case for them. But either way, my syntax proposal for what to launch now is the same (see below).

## Should every element be a 'style container' by default, since no containment is needed?

We initially said 'yes', because this entire feature is designed around authors getting the most immediate available information by default. For usability of the feature, this is the right way to go. Container-name is a better solution than container-type for filtering the list of containers when necessary from an author perspective. It is also likely to make sense down the road for other hypothetical query-types that don't rely on specialized containment, since it seems like the ancestor-descendant requirement is enough to protect most query-types from looping.

### Are there performance issues that should make us reconsider that?

The answer depends on both the implementation approach, and the way authors write their queries. Those potential performance issues only apply to the resolution of style queries, they are not a general performance penalty for all sites across the board.

Container queries involve two levels of element 'filtering' - one to find the 'containers' elements that will be queried, and the other to find the selector 'target' elements impacted by the query. Currently, the Gecko and Chromium implementations filter selector targets first, and then use that information to narrow the search for viable containers (which must be an ancestor of the target). The Webkit implementation takes the reverse approach - first filtering viable containers, and then selector-matching only within their descendant trees. As with the internals of selector-matching, the more narrowly targeted the first step, the better the performance. If selector-matching narrows the list farther, then it ought to go first, and vice-versa.

If all elements are style containers (either by default, or through a global declaration), then a style query without a container-name is very broadly targeting all elements as potential containers. That will impact the performance of all container-first filtering algorithms. The first step has basically no filtering impact. For selector-first algorithm, the impact is relative to the specificity of selectors involved. In either case, it is possible for authors to cause performance issues through a combination of broad selectors inside broadly-targeted queries. The concern is that having more containers on the page makes more queries 'broadly-answerable'.

Since authors are likely to apply this global rule if we don't, it seems essential that we optimize for it, no matter how we conclude this issue. It's also possible that selector-first filtering will have some advantages here. In the meantime, authors can avoid those performance concerns by using container-names to narrow the container list, or keep selectors more narrowly targeted to avoid broadly-applied queries.

However, it's still not clear to me _how much_ of a performance slow-down we're talking about. While the `*` selector is less performant than an `#id`, that hasn't been enough of a difference to impact most developers choice of selectors.

## Can we help authors avoid some container declarations overriding others?

With both container-types and container-names, it's likely that authors will want to apply different values in different places – and ideally would be able to merge those declarations, rather than one overriding another. That's an issue that has come up with many other CSS properties, and is the driving concern behind the idea of 'additive cascade'. 

In this case, we could consider breaking out individual longhand properties for different categories of container-type (size, style, etc) which would allow those types to be set in parallel. But: 

- That doesn't help with conflicts inside a given category (size vs inline-size)
- It only works if authors entirely avoid the shorthand, which would reset all longhand values anyway
- The conflict is even more likely with container-names, which cannot be broken into shorthands

I don't think the benefits here are worth the effort. I would rather see us work on a general solution to the problem of additive cascade.

## My proposal

These are all good concerns to raise, and I'm glad to be walking through them in some detail. I find it very hard to weigh the performance concerns in relation to author usability - since it involves a number of factors that change over time. Clearly, implementors have a better sense of that than I do. But at this point I still feel that we had the right proposal here initially.

I would like to resolve that:

- The initial value of `container-type` is `none`.
- All elements are `style` containers, no matter what `container-type` value is applied.
- Browsers should feel free to ship size queries as they are implemented, and ship style queries later.

In the meantime, I'll work on providing more detailed use-cases for style queries, so that the reason for implementing that feature is clear. Even if I fail at that, this resolution on the initial value of container-type allows us to ship the best syntax for size queries now. And if anyone wants to work on a proposal for additive cascade, I would love to help push that forward.

**Comment by matthew-dean** [comment]
@mirisuzanne 

>  if anyone wants to work on a proposal for additive cascade, I would love to help push that forward.

`* cracks knuckles *`

**Comment by bramus** [comment]
For completeness, linking to https://github.com/w3c/csswg-drafts/issues/1594 here. It’s the issue talking about “Additive Cascade”, a concept which was mentioned a lot here.
